contracts/MerkleDistributorWithDeadline.sol

```diff
@@ -0,0 +1,31 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity =0.8.15;
+
+import {MerkleDistributor} from "./MerkleDistributor.sol";
+import {Ownable} from "@openzeppelin/contracts/access/Ownable.sol";
+import {IERC20, SafeERC20} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+error EndTimeInPast();
+error ClaimWindowFinished();
+error NoWithdrawDuringClaim();
+
+contract MerkleDistributorWithDeadline is MerkleDistributor, Ownable {
+    using SafeERC20 for IERC20;
+
+    uint256 public immutable endTime;
+
+    constructor(address token_, bytes32 merkleRoot_, uint256 endTime_) MerkleDistributor (token_, merkleRoot_) {
+        if (endTime_ <= block.timestamp) revert EndTimeInPast();
+        endTime = endTime_;
+    }
+
+    function claim(uint256 index, address account, uint256 amount, bytes32[] calldata merkleProof) public override {
+        if (block.timestamp > endTime) revert ClaimWindowFinished();
+        super.claim(index, account, amount, merkleProof);
```

Comment:
maybe instead of inheriting MerkleDistributor and using super thereby copying the bytes32[] array into memory using a private method in a commonly inherited BaseContract would be preferable to preserve the external visibility